README
NAME
    WWW::Finger - get useful data from e-mail addresses

VERSION
    0.09

SYNOPSIS
      use WWW::Finger;
      my $finger = WWW::Finger->new("joe@example.com");
      if (defined $finger)
      {
        print $finger->name . "\n";
      }

DESCRIPTION
    This module is *not* an implementation of the finger protocol (RFC
    1288). Use Net::Finger for that. Instead it is a set of implementations
    of *other* methods for getting information from an e-mail address, or
    e-mail-like identifier. This package includes four such implementations,
    and it's pretty easy to create your own additional implementations:

    *       WebFinger

    *       Fingerpoint

    *       cpan.org scraper (for user@cpan.org)

    *       Unnamed finger protocol described on bitworking.org

    Only the first two implementations are enabled by default. See
    "IMPLEMENTATIONS" on how to enable others, and for more details.

  Constructor
    *       "new"

              $finger = WWW::Finger->new($identifier);

            Creates a WWW::Finger object for a particular identifier. Will
            return undef if no implemetation is able to handle the
            identifier

  Object Methods
    Any of these methods can return undef if the appropriate information is
    not available. The "name", "mbox", "homepage", "weblog", "image" and
    "key" methods work in both scalar and list context. Depending on which
    implementation was used by "WWW::Finger::new", the object may also have
    additional methods. Consult the documentation of the various
    implementations for details.

    "name"  The person's name (or handle/nickname).

    "mbox"  The person's e-mail address (including "mailto:").

    "homepage"
            The person's personal homepage.

    "weblog"
            The person's blog. (There may be some overlap with "homepage".)

    "image" An avatar, photo or other image depicting the person.

    "key"   The URL of the person's GPG/PGP public key.

    "webid" A URI uniquely identifying the person. See
            <http://esw.w3.org/topic/WebID>.

    "endpoint"
            A SPARQL Protocol endpoint which may provide additional data
            about the person. (See RDF::Query::Client.)

    "graph" An RDF::Trine::Model object holding data about the person. (See
            RDF::Trine.)

IMPLEMENTATIONS
  Loading Additional Implementations
    When importing this package ("use WWW::Finger") you can pass it a list
    of additional finger implementations to load. For example:

      use WWW::Finger qw(WWW::Finger::CPAN MyCorp::Finger);

    For packages which start with "WWW::Finger::" there is a special
    abbreviation:

      use WWW::Finger qw(+CPAN MyCorp::Finger);

    Assuming the finger implementations are written correctly,
    WWW::Finger->new should just notice they exist and use them when
    appropriate.

    The WebFinger and Fingerpoint implementations are loaded by default.

    To load additional implementations later on (after you've already "used"
    WWW::Finger) you can call the import method:

      WWW::Finger->import(qw(+Foo +Bar MyCorp::Baz));

    There's no official method of removing an already-imported
    implementation, but if you really need to, try playing around with the
    @WWW::Finger::Modules array.

  Calling an Implementation Specifically
    If you need to call a particular implementation specifically, that
    should be fairly simple:

      use WWW::Finger::WebFinger;
      my $finger = WWW::Finger::WebFinger->new("joe@example.com");
      if (defined $finger)
      {
        print $finger->name . "\n";
      }

  Writing Your Own Implementation
    Use this stub:

      package WWW::Finger::Example;
  
      use strict;
      use WWW::Finger;
      use URI;
  
      our @ISA = qw(WWW::Finger);
      our $VERSION = '0.01';
  
      BEGIN { push @WWW::Finger::Modules, __PACKAGE__; }
  
      sub new
      {
        my ($class, $identifier) = @_;
        my $self = {};
    
        # Canonicalise the identifier. You don't have to use
        # "mailto:"; other URI schemes are allowable.
        $identifier = "mailto:$identifier"
          unless $identifier =~ /^[a-z0-9\.\-\+]+:/i;

        # Check whether this package can get useful info
        # from $identifier. If not, then return undef.
        if ('check things here')
        {
          return undef;
        }
    
        $self->{'ident'} = URI->new($identifier);
        bless $self, $class;
      }
  
      # Override WWW::Finger methods like 'name', 'mbox', etc.
      # Feel free to provide additional methods too.
  
      1;

SEE ALSO
    Net::Finger.

    <http://code.google.com/p/webfinger/>.

    <http://buzzword.org.uk/2009/fingerpoint/spec>.

    <http://www.perlrdf.org/>.

    fingerw.

AUTHOR
    Toby Inkster, <tobyink@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2009-2010 by Toby Inkster

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8 or, at your
    option, any later version of Perl 5 you may have available.



Changes
############################################################################
## Changes #################################################################
############################################################################

WWW-Finger
==========

Created:      2009-12-08
Home page:    <http://search.cpan.org/dist/WWW-Finger/>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=WWW-Finger>
Maintainer:   Toby Inkster <mailto:tobyink@cpan.org>

0.101 [2010-07-31]
 - (Addition) Webfinger and Fingerpoint modules provide a 'get' method to
   easily access additional data.

0.100 [2010-07-25]
 - (Addition) CPAN module now has webid method.
 - (Update) Use the somewhat more complete (albeit slower) HTTP::LRDD
   module to perform LRDD for Webfinger rather than a home-grown method.

0.09 [2010-04-26]
 - Various cleanups.
 - (Update) Switch Changes file from Aaron Cope's changefile vocab to my
   own DOAP Change Sets vocab.

0.08 [2010-01-06]
 - (Bugfix) Documentation improvements.

0.07 [2009-01-06]
 - (Bugfix) Include BitworkingFingerProtocol.pm in MANIFEST.
 - (Update) Update copyright noticed to 2010.

0.06 [2009-01-06]
 - Refactor some code from WWW::Finger::Webfinger to allow it to be used
   elsewhere.
 - (Addition) Allow fingerw command-line client to load additional modules.
 - (Addition) Follow rdf:seeAlso and ianalink:describedby links found in
   Webfinger account profiles.
 - (Addition) Support Joe Gregorio's alternative to current Webfinger
   protocol.
 - (Addition) Support semantics of foaf:account and foaf:holdsAccount.
 - (Bugfix) Fix duplicated mbox in WWW::Finger::CPAN.
 - (Update) Improve fingerw's debugging (verbose) messages.
 - (Update) Use Turtle for Changes file.

0.05 [2009-12-16]
 - (Bugfix) Mini bug-fixes.
 - (Update) use XRD::Parser 0.04.

0.04 [2009-12-16]
 - Change default scheme for Webfinger from 'acct://' to 'acct:'.
 - (Addition) Bundle a command-line query tool.
 - (Bugfix) Fingerpoint bugfix in ->graph().
 - (Bugfix) Unimplemented methods return empty arrays properly.

0.03 [2009-12-15]
 - (Addition) WWW::Finger::Webfinger now provides an ->endpoint method.
 - (Bugfix) Fixed package signature.

0.02 [2009-12-14]
 - (Bugfix) Fixed problems with build.

0.01 [2009-12-14]



lib/WWW/Finger/Fingerpoint.pm
package WWW::Finger::Fingerpoint;

use 5.008;
use base qw[WWW::Finger];
use common::sense;

use Carp;
use Digest::SHA1 qw[sha1_hex];
use HTTP::Link::Parser qw[:standard];
use LWP::UserAgent;
use RDF::Query::Client;
use RDF::Trine;
use URI;
use WWW::Finger;

our $VERSION = '0.101';

my $rel_fingerpoint = 'http://ontologi.es/sparql#fingerpoint';

BEGIN
{
	push @WWW::Finger::Modules, __PACKAGE__;
}

sub new
{
	my $class = shift;
	my $ident = shift or croak "Need to supply an e-mail address\n";
	my $self  = bless {}, $class;
		
	$ident = "mailto:$ident"
		unless $ident =~ /^[a-z0-9\.\-\+]+:/i;
	$ident = URI->new($ident);
	return undef
		unless $ident->scheme eq 'mailto';
	
	$self->{'ident'} = $ident;
	my ($user, $host) = split /\@/, $ident->to;
	
	my $ua = LWP::UserAgent->new;
	$ua->timeout(10);
	$ua->env_proxy;
	
	my $httphost = "http://$host/";
	my $response = $ua->head($httphost);
	return undef
		unless $response->is_success;
	
	my $linkdata = HTTP::Link::Parser::parse_links_to_rdfjson($response);
	my $sparql   = $linkdata->{ $httphost }->{ $rel_fingerpoint }->[0]->{'value'};

	unless (defined $sparql)
	{
		$response = $ua->get($httphost,
			'Accept' => 'application/xhtml+xml;q=1.0, text/html;q=0.9, */*;q=0.1');
		return undef
			unless $response->is_success;
		if ($response->header('content-type') =~ m`^(text/html|application/xhtml+xml|application/xml|text/xml)`i)
		{
			$sparql = URI->new_abs($1, URI->new($httphost))
				if $response->content =~ m`<[Ll][Ii][Nn][Kk]\s+[Rr][Ee][Ll]="[^"]*http://ontologi\.es/sparql#fingerpoint[^"]*"\s+[Hh][Rr][Ee][Ff]="([^"]+)"\s*/?>`;
		}
	}

	return undef
		unless defined $sparql && length $sparql;
	
	$self->{'endpoint'} = $sparql;
	
	my $sha1 = sha1_hex($ident);
	my $sparql_query = "PREFIX foaf: <http://xmlns.com/foaf/0.1/>
	PREFIX wot: <http://xmlns.com/wot/0.1/>
	SELECT DISTINCT *
	WHERE {
		{
			{ ?person foaf:mbox <$ident> . }
			UNION
			{ ?person foaf:mbox_sha1sum \"$sha1\" . }
			UNION
			{ ?person foaf:account <$ident> . }
			UNION
			{ ?person foaf:holdsAccount <$ident> . }
		}
		OPTIONAL { ?person foaf:name ?name . }
		OPTIONAL { ?person foaf:nick ?nick . }
		OPTIONAL { ?person foaf:homepage ?homepage . }
		OPTIONAL { ?person foaf:mbox ?mbox . }
		OPTIONAL { ?person foaf:weblog ?weblog . }
		OPTIONAL { ?person foaf:img ?image . }
		OPTIONAL { ?k wot:pubkeyAddress ?key ; wot:identity ?person . }
	}";
	my @fields = qw(name homepage mbox weblog image key);
	
	my $query  = RDF::Query::Client->new($sparql_query);
	my $result = $query->execute($self->endpoint, {QueryMethod=>'POST'});
	my $webid;

	BINDING: while (my $binding = $result->next)
	{
		$webid = $binding->{'person'}->uri
			if  $binding->{'person'}
			and $binding->{'person'}->is_resource
			and !defined $webid;
			
		FIELD: foreach my $field (@fields)
		{
			next FIELD unless $binding->{$field};
			
			if ($binding->{$field}->is_resource)
				{ $self->{'data'}->{$field}->{ $binding->{$field}->uri } = 1; }
			elsif ($binding->{$field}->is_literal)
				{ $self->{'data'}->{$field}->{ $binding->{$field}->literal_value } = 1; }
		}
	}
	
	foreach my $field (@fields)
	{
		$self->{'data'}->{$field} = [ keys %{ $self->{'data'}->{$field} } ];
	}
	
	$self->{'webid'} = $webid;
	
	return $self;
}

sub graph
{
	my $self = shift;
	
	unless (defined $self->{'graph'})
	{
		my $ident = $self->{'ident'}.'';
		my $sha1 = sha1_hex($ident);
		my $model = RDF::Trine::Model->new( RDF::Trine::Store->temporary_store );
		my $query  = RDF::Query::Client->new("
			PREFIX foaf: <http://xmlns.com/foaf/0.1/>
			DESCRIBE ?person
			WHERE
			{
				{ ?person foaf:mbox <$ident> . }
				UNION
				{ ?person foaf:mbox_sha1sum \"$sha1\" . }
				UNION
				{ ?person foaf:account <$ident> . }
				UNION
				{ ?person foaf:holdsAccount <$ident> . }
			}");
		my $result = $query->execute($self->endpoint, {QueryMethod=>'POST'});
		if ($result)
		{
			while (my $st = $result->next)
			{
				$model->add_statement($st);
			}
			$self->{'graph'} = $model;
		}		
	}
	
	return $self->{'graph'};
}

sub get
{
	my ($self, @params) = @_;
	return WWW::Finger::_GenericRDF::_simple_sparql(
		$self,
		{use_endpoint=>1},
		map { HTTP::Link::Parser::relationship_uri($_) } @params );
}

sub endpoint
{
	my $self = shift;
	return $self->{'endpoint'};
}

sub webid
{
	my $self = shift;
	return $self->{'webid'};
}

sub _data
{
	my $self = shift;
	my $k    = shift;
	if (wantarray)
	{
		return @{ $self->{'data'}->{$k} }
			if defined $self->{'data'}->{$k};
	}
	else
	{
		return $self->{'data'}->{$k}->[0]
			if defined $self->{'data'}->{$k}->[0];
	}
	return undef;
}

sub name
{
	my $self = shift;
	return $self->_data('name');
}

sub nick
{
	my $self = shift;
	return $self->_data('nick');
}

sub mbox
{
	my $self = shift;
	return $self->_data('mbox');
}

sub image
{
	my $self = shift;
	return $self->_data('image');
}

sub homepage
{
	my $self = shift;
	return $self->_data('homepage');
}

sub weblog
{
	my $self = shift;
	return $self->_data('weblog');
}

sub key
{
	my $self = shift;
	return $self->_data('key');
}

1;
__END__

=head1 NAME

WWW::Finger::Fingerpoint - Investigate E-mail Addresses using Fingerpoint

=head1 SYNOPSIS

  ## Using WWW::Finger
  
  use WWW::Finger;
  
  my $finger = WWW::Finger->new("joe@example.com");
  
  if ($finger)
  {
    if ($finger->isa('WWW::Finger::Fingerpoint'))
    {
      print "WWW::Finger used WWW::Fingerpoint\n";
    }
    print $finger->name . "\n";  # print person's name.
 }

  ## Using WWW::Finger::Fingerpoint directly
  
  use RDF::Query::Client;
  use WWW::Finger::Fingerpoint;
  
  my $fingerpoint = WWW::Finger::Fingerpoint->new("joe@example.com");
  
  if ($fingerpoint->webid)
  {
    my $sparql  = sprintf(
      "SELECT * WHERE {<%s> <http://xmlns.com/foaf/0.1/homepage> ?page.}",
      $fingerpoint->webid);
    my $query   = RDF::Query::Client->new($sparql);
    my $results = $query->execute($fingerpoint->endpoint);
	 while (my $row = $results->next)
    {
      print "Found page: " . $row->{'page'}->uri . "\n";
    }
  }
  
=head1 DESCRIPTION

As well as the standard WWW::Finger methods, WWW::Finger::Fingerpoint provides this
additional method:

=over 4

=item C<< get($p1, $p2, ...) >>

$p1, $p2 and are RDF predicate URIs. Returns a list of values which are non-bnode
objects of triples where the predicate URI is one of the parameters and the 
subject URI is the person/agent fingered.

  # Returns phone numbers...
  $finger->get('http://xmlns.com/foaf/0.1/phone',
               'http://rdf.data-vocabulary.org/#tel');

=back

=head1 SEE ALSO

L<WWW::Finger>.

L<RDF::Query::Client>, L<RDF::Trine>.

L<http://buzzword.org.uk/2009/fingerpoint/spec>.

L<http://www.perlrdf.org/>.

=head1 AUTHOR

Toby Inkster, E<lt>tobyink@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009-2010 by Toby Inkster

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
